### ISSUE TITLE
Support for unity build with il2cpp

### ISSUE THREAD

**Issue by elfasito** [EPPlusSoftware/EPPlus#702]
Hello people, I have downloaded epplus 6.0.6 from nuget and imported to my test project without problems. (works fine on windows)
The issue appears after build on android device.
I get this error:
```
NotSupportedException: System.Configuration.ConfigurationManager::get_AppSettings
System.Configuration.ConfigurationManager.get_AppSettings () [0x00000] in <00000000000000000000000000000000>:0
```
I checked on unity forums and effectively IL2CPP scripting dont have support for System.ConfigurationManager.
Is there a way to workaround this?, I appreciate any guidance, thanks.

**Comment by swmal** [comment]
Hello,
is there a more detailed stack trace of your Exception so we can see where in EPPlus this Exception is thrown? We are currently working on error handling for the calls to Configuration end Environment variables so it won't crash. We hope to launch this in the next revision of EPPlus.

**Comment by elfasito** [comment]
> Hello, is there a more detailed stack trace of your Exception so we can see where in EPPlus this Exception is thrown? We are currently working on error handling for the calls to Configuration end Environment variables so it won't crash. We hope to launch this in the next revision of EPPlus.

Hi swmal, Unity player dont give me debug about the dll of epplus, Im readed for a while how achieve it but no lucky (im novice programer, sorry).
I only can see what the error happens at this line of my code:
```
> using (ExcelPackage excelPackage = new ExcelPackage(file))
            {
             ExcelWorksheet worksheet = excelPackage.Workbook.Worksheets["Hoja1"];
              ...........
            }
on visual studio attached to my unity player get this:
```
System.NotSupportedException: 
System.Configuration.ConfigurationManager::get_AppSettings
```

on dalvikvm debug this:
```
 NotSupportedException: System.Configuration.ConfigurationManager::get_AppSettings
08-26 17:26:07.079 26347 26365 E Unity   :   at System.Configuration.ConfigurationManager.get_AppSettings () [0x00000] in <00000000000000000000000000000000>:0
08-26 17:26:07.079 26347 26365 E Unity   :   at OfficeOpenXml.ExcelPackage.Init () [0x00000] in <00000000000000000000000000000000>:0
08-26 17:26:07.079 26347 26365 E Unity   :   at OfficeOpenXml.ExcelPackage..ctor (System.IO.FileInfo newFile) [0x00000] in <00000000000000000000000000000000>:0
```

**Comment by swmal** [comment]
You can try our development feed and see if that resolved the issue:

https://github.com/EPPlusSoftware/EPPlus/wiki/Using-our-Develop-Nuget-Feed

Before creating an instance of ExcelPackage you must call the new static Configure method like this (C#):

```csharp
ExcelPackage.LicenseContext = LicenseContext.NonCommercial;
ExcelPackage.Configure(x => x.SuppressInitializationExceptions = true);
using var package = new ExcelPackage()
```

I assume this lambda becomes a delegate in C++?

**Comment by elfasito** [comment]
> 

Hi swmal I tried with the new dev update.
I get this on dalvikvm debug:
```
 MissingMemberException: The lazily-initialized type does not have a public, parameterless constructor.
08-29 19:09:22.902 16822 16840 E Unity   :   at System.Threading.LazyHelpers`1[T].ActivatorFactorySelector () [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at System.Threading.LazyInitializer.EnsureInitializedCore[T] (T& target, System.Boolean& initialized, System.Object& syncLock, System.Func`1[TResult] valueFactory) [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at System.Threading.LazyInitializer.EnsureInitialized[T] (T& target, System.Boolean& initialized, System.Object& syncLock) [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at OfficeOpenXml.Utils.RecyclableMemory.get_MemoryManager () [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at OfficeOpenXml.Utils.RecyclableMemory.GetStream () [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at OfficeOpenXml.ExcelPackage.ConstructNewFile (System.String password) [0x00000] in <00000000000000000000000000000000>:0
08-29 19:09:22.902 16822 16840 E Unity   :   at OfficeOpenXml.ExcelPackage..ctor (System.IO.FileInfo newFil
```

and this on visual studio debug:
```
> using (ExcelPackage excelPackage = new ExcelPackage(file)) 
System.MissingMethodException: Default constructor not found for type Microsoft.IO.ReciclableMemoryStreamManager
```

**Comment by swmal** [comment]
Hi @elfasito,

This seems a bit strange, because the `Microsoft.IO.RecyclableMemoryStreamManager` class that is initialized by the [System.Threading.LazyInitializer](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=net-6.0) do have a parameterless constructor. See [the code](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream/blob/master/src/RecyclableMemoryStreamManager.cs#L87) of this class. The debug output you have provided above says Microsoft.IO.**Reciclable**MemoryStreamManager, i.e. Recyclable is misspelled. Is that what your really get from the debugger or is it just a typo?

We could look at replacing the usage of the LazyInitializer with a standard singleton implementation, but it is still odd that it this works in all other environments.

**Comment by elfasito** [comment]
> Hi @elfasito,
> 
> This seems a bit strange, because the `Microsoft.IO.RecyclableMemoryStreamManager` class that is initialized by the [System.Threading.LazyInitializer](https://docs.microsoft.com/en-us/dotnet/api/system.threading.lazyinitializer.ensureinitialized?view=net-6.0) do have a parameterless constructor. See [the code](https://github.com/microsoft/Microsoft.IO.RecyclableMemoryStream/blob/master/src/RecyclableMemoryStreamManager.cs#L87) of this class. The debug output you have provided above says Microsoft.IO.**Reciclable**MemoryStreamManager, i.e. Recyclable is misspelled. Is that what your really get from the debugger or is it just a typo?
> 
> We could look at replacing the usage of the LazyInitializer with a standard singleton implementation, but it is still odd that it this works in all other environments.

Hi swmal.
Its a copy&paste from the debugger, I did not write nothing by myself.

**Comment by swmal** [comment]
Hi @elfasito,

You can try again using the latest version in our Development feed for EPPlus 6 (should have been built and published some 20 minutes after this post). I have added Exception handling and fallback functionality when initializing this class. 

Let me know if this resolves your problem.

**Comment by swmal** [comment]
Ok, that was one step ahead...

This seems to be related to android somehow, found [this thread](https://github.com/adamhathcock/sharpcompress/issues/405)

We'll have a look at it to see if there is anything we can do on our side.

**Comment by swmal** [comment]
Try again with the latest version in the development feed, we have removed the code that tried to fetch this Encoding.

**Comment by illmaticGamedev** [comment]
@swmal hey,
I am trying to build on windows with the latest version but im still seeing this error, any help with that ?

![image](https://user-images.githubusercontent.com/80155331/187920460-e4fd7b1b-58f2-4009-9958-d6fc3efd3d89.png)

**Comment by JanKallman** [comment]
Are you using 6.0.7.1044-20220901-develop? I have not added this to 5.8.13 yet.

**Comment by illmaticGamedev** [comment]
I simply downloaded the package from nuget gallary
https://www.nuget.org/packages/EPPlus

**Comment by elfasito** [comment]
> Try again with the latest version in the development feed, we have removed the code that tried to fetch this Encoding.

Hi @swmal, I updated to the lastest develop.
I get this debug on visual studio:
```
  System.TypeInitializationException: The type initializer for 'OfficeOpenXml.Packaging.Ionic.Zip.SharedUtilities' threw an exception. ---> System.NotSupportedException: Encoding 437 data could not be found. Make sure you have correct international codeset assembly installed and enabled.
  at System.Text.Encoding.GetEncoding (System.Int32 codepage) [0x00000] in <00000000000000000000000000000000>:0 
  at System.Text.Encoding.GetEncoding (System.String name) [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.Packaging.Ionic.Zip.SharedUtilities..cctor () [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.Packaging.Ionic.Zip.ZipInputStream.GetNextEntry () [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.Packaging.ZipPackage..ctor (System.IO.Stream stream) [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.ExcelPackage.ConstructNewFile (System.String password) [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.ExcelPackage..ctor (System.IO.FileInfo newFile) [0x00000] in <00000000000000000000000000000000>:0 
  at WRITEXLSLIGHT.TaskWriteXLSX () [0x00000] in <00000000000000000000000000000000>:0 
  at WRITEXLSLIGHT.GenerateEmptyfile () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Events.InvokableCall.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Events.UnityEvent.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UI.Button.Press () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UI.Button.OnPointerClick (UnityEngine.EventSystems.PointerEventData eventData) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.ExecuteEvents.Execute (UnityEngine.EventSystems.IPointerClickHandler handler, UnityEngine.EventSystems.BaseEventData eventData) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.ExecuteEvents.Execute[T] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.ExecuteEvents+EventFunction`1[T1] functor) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchPress (UnityEngine.EventSystems.PointerEventData pointerEvent, System.Boolean pressed, System.Boolean released) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchEvents () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.Process () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.EventSystem.Update () [0x00000] in <00000000000000000000000000000000>:0 
   --- End of inner exception stack trace ---
  at OfficeOpenXml.Packaging.Ionic.Zip.ZipInputStream.GetNextEntry () [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.Packaging.ZipPackage..ctor (System.IO.Stream stream) [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.ExcelPackage.ConstructNewFile (System.String password) [0x00000] in <00000000000000000000000000000000>:0 
  at OfficeOpenXml.ExcelPackage..ctor (System.IO.FileInfo newFile) [0x00000] in <00000000000000000000000000000000>:0 
  at WRITEXLSLIGHT.TaskWriteXLSX () [0x00000] in <00000000000000000000000000000000>:0 
  at WRITEXLSLIGHT.GenerateEmptyfile () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Events.InvokableCall.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.Events.UnityEvent.Invoke () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UI.Button.Press () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.UI.Button.OnPointerClick (UnityEngine.EventSystems.PointerEventData eventData) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.ExecuteEvents.Execute (UnityEngine.EventSystems.IPointerClickHandler handler, UnityEngine.EventSystems.BaseEventData eventData) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.ExecuteEvents.Execute[T] (UnityEngine.GameObject target, UnityEngine.EventSystems.BaseEventData eventData, UnityEngine.EventSystems.ExecuteEvents+EventFunction`1[T1] functor) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchPress (UnityEngine.EventSystems.PointerEventData pointerEvent, System.Boolean pressed, System.Boolean released) [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.ProcessTouchEvents () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.StandaloneInputModule.Process () [0x00000] in <00000000000000000000000000000000>:0 
  at UnityEngine.EventSystems.EventSystem.Update () [0x00000] in <00000000000000000000000000000000>:0 
```

I make another try with the last develop version and I added these libraries to my project: I18N.dll & I18N.West.dll (recommended in the thread you shared)
and now its working.

**Comment by swmal** [comment]
No, that is not related to the previous problems. Would be great if you could upload a sample png-image that generates this error and we will try to fix it.

We'll have another look at the stacktrace you provided, the 437 Encoding isn't really needed by EPPlus but included in the code we use for managing zip-streams.

**Comment by illmaticGamedev** [comment]
Hey @swmal is there a timeline on when the last update will be update for the nuget package ?

**Comment by elfasito** [comment]
> Hey @swmal is there a timeline on when the last update will be updated for the nuget package ?

Hi @swmal
this is a example of a png what im trying to add (its generated on runtime)
http://imgfz.com/i/he7ZMGY.png

and I using this code to add it (not working only on android):
```
 if (File.Exists(imgPath))
                {
                    byte[] imgFile = File.ReadAllBytes(imgPath);
                    imgFile = File.ReadAllBytes(imgPath);
                    ExcelPicture pic = worksheet.Drawings.AddPicture("Sign", imgPath);
                    pic.SetPosition(85, 8, 4, 13); //format: (1,2,3,4) > 1:row number, 3:column number
                    pic.SetSize(225, 145);
                }
```

**Comment by elfasito** [comment]
> No, that is not related to the previous problems. Would be great if you could upload a sample png-image that generates this error and we will try to fix it.
> 
> We'll have another look at the stacktrace you provided, the 437 Encoding isn't really needed by EPPlus but included in the code we use for managing zip-streams.

> Hey @swmal is there a timeline on when the last update will be updated for the nuget package ?

Hi @swmal
this is a example of a png what im trying to add (its generated on runtime)
http://imgfz.com/i/he7ZMGY.png

and I using this code to add it (not working only on android):
```
 if (File.Exists(imgPath))
                {
                    byte[] imgFile = File.ReadAllBytes(imgPath);
                    imgFile = File.ReadAllBytes(imgPath);
                    ExcelPicture pic = worksheet.Drawings.AddPicture("Sign", imgPath);
                    pic.SetPosition(85, 8, 4, 13); //format: (1,2,3,4) > 1:row number, 3:column number
                    pic.SetSize(225, 145);
                }
```

**Comment by swmal** [comment]
@illmaticGamedev I cannot give you an exact date, but it will most likely be within the next two weeks. You can use our development nuget feed for dev purposes in the meantime.

**Comment by JanKallman** [comment]
@elfasito The image attached, he7ZMGY.png, is actually a jpg. Rename it *. jpg or add it using a stream with the option `ePictureType.Jpg`.

**Comment by elfasito** [comment]
> @elfasito The image attached, he7ZMGY.png, is actually a jpg. Rename it *. jpg or add it using a stream with the option `ePictureType.Jpg`.

Hi @JanKallman 
the original image was a PNG but the server I used to save it modify the image (I just notice now)
I share a new link: https://ibb.co/VLsyKX9
at least on notepad looks like a png file.
Anyways I tried to rename the image to jpg but I get the same error.

**Comment by JanKallman** [comment]
@elfasito Thanks, Oddly this issue was just fixed yesterday, see #711. You can try it using latest develop build... https://github.com/EPPlusSoftware/EPPlus/wiki/Using-our-Develop-Nuget-Feed

**Comment by elfasito** [comment]
> 

Hi @JanKallman 
I updated to: 6.0.7.1048-20220906-develop
but still Im having the same problem handling png images.
```
09-07 16:35:32.433 29127 29145 E Unity   : InvalidOperationException: No image handler for image type Png
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.PictureStore.GetImageBounds (System.Byte[] image, OfficeOpenXml.Drawing.ePictureType type, OfficeOpenXml.ExcelPackage pck) [0x00000] in <00000000000000000000000000000000>:0
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.PictureStore.AddImage (System.Byte[] image, System.Uri uri, System.Nullable`1[T] pictureType) [0x00000] in <00000000000000000000000000000000>:0
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.ExcelPicture.SaveImageToPackage (OfficeOpenXml.Drawing.ePictureType type, System.Byte[] img) [0x00000] in <00000000000000000000000000000000>:0
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.ExcelPicture.LoadImage (System.IO.Stream stream, OfficeOpenXml.Drawing.ePictureType type) [0x00000] in <00000000000000000000000000000000>:0
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.ExcelDrawings.AddPicture (System.String Name, System.IO.FileInfo ImageFile, System.Uri Hyperlink) [0x00000] in <00000000000000000000000000000000>:0
09-07 16:35:32.433 29127 29145 E Unity   :   at OfficeOpenXml.Drawing.ExcelDrawings.AddPicture (System.String Name,
```

**Comment by swmal** [comment]
All fixes that are related to this issue is 6.0.8 (released on Nuget today). If you discover further issues please open a new issue.
